### PR TITLE
[FIX] web: compute containing block layout in position hook

### PR DESCRIPTION
**Pull request by abd-msyukyu-odoo**
[FIX] web: compute containing block layout in position hook
How to reproduce:
- create a list embedded view in Knowledge (/item list)
- in the control panel click on "Filters"

Current behavior:
- the dropdown is not correctly positioned

Expected behavior:
- the dropdown should be positioned around the button

Fix:
`getBestPosition` uses dimensions and layout positionning based on the viewport,
but `reposition` modifies the style properties `top` and `left` of the element,
which are relative to the closest containing block. Since in that case the style
position is `fixed`, there are a number of factor that can influence the
containing block, as specified in:
https://developer.mozilla.org/en-US/docs/Web/CSS/Containing_block#identifying_the_containing_block
We can easily compute the layout offset caused by the containing block just
after resetting the popper top and left to `0px` and take it into account for
the final value.

Task-3267555

### CODE REVIEW

File: addons/web/static/tests/core/position_hook_tests.js [hunk 1]
```diff
@@ -150,6 +150,42 @@ QUnit.test("has no effect when component is destroyed", async (assert) => {
     );
 });
 
+QUnit.test("is positioned relative to its containing block", async (assert) => {
+    assert.expect(4);
```
Comment (brboi) [inline addons/web/static/tests/core/position_hook_tests.js:154]: not needed, we try to only specify `assert.expect(x)` if the number of assertion could be variable (like asserting inside a patched function or the makeView util mockRPC param.

File: addons/web/static/src/core/position_hook.js [hunk 2]
```diff
@@ -101,6 +103,12 @@ function getBestPosition(reference, popper, { container, margin, position }) {
     const refBox = reference.getBoundingClientRect();
     const contBox = container.getBoundingClientRect();
 
+    // Get the offsets of the containing block (relative to the viewport)
+    // https://developer.mozilla.org/en-US/docs/Web/CSS/Containing_block#identifying_the_containing_block
+    const popStyle = window.getComputedStyle(popper);
```
Comment (brboi) [inline addons/web/static/src/core/position_hook.js:108]: I don't understand why is this needed ?
We call this `getBestPosition` function right after setting the top and left style properties to "0px".
Can you explain ?

File: addons/web/static/src/core/position_hook.js [hunk 3]
```diff
@@ -101,6 +103,12 @@ function getBestPosition(reference, popper, { container, margin, position }) {
...
+    const verticalOffset = (popBox.top - parseFloat(popStyle.top)) || 0;
+    const horizontalOffset = (popBox.left - parseFloat(popStyle.left)) || 0;
```
Comment (brboi) [inline addons/web/static/src/core/position_hook.js:110]: If I am not wrong, as we already know the popper style at this point, we could simply do nothing here:
```suggestion
```

And below, in getPositioningData:
```js
return {
    top: positioning.top - popBox.top,
    left: positioning.left - popBox.left,
    direction: DIRECTIONS[d],
    variant: VARIANTS[v],
};
```

File: addons/web/static/tests/core/position_hook_tests.js [hunk 4]
```diff
@@ -150,6 +150,42 @@ QUnit.test("has no effect when component is destroyed", async (assert) => {
...
+    const popBox2 = document.getElementById("popper").getBoundingClientRect();
+    destroy(popper);
+
+    // best positions are not the same relative to their containing block
+    assert.notEqual(pos1.top, pos2.top);
+    assert.notEqual(pos1.left, pos2.left);
+    // best positions are the same relative to the viewport
+    assert.equal(popBox1.top, popBox2.top);
+    assert.equal(popBox1.left, popBox2.left);
+});
```
Comment (brboi) [inline addons/web/static/tests/core/position_hook_tests.js:187]: What do you think about avoiding `assert.notEqual` (which I don't like as it is not precise) with something similar to the following (I did not try) ?
```suggestion
    const margin = 15;
    container.style.margin = `${margin}px`;
    const positions = [];
    const TestComp = getTestComponent({
        onPositioned: (el, pos) => {
            positions.push(pos);
        },
    });
    let popper = await mount(TestComp, container);
    const popBox1 = document.getElementById("popper").getBoundingClientRect();
    destroy(popper);
    document.getElementById("reference").remove();

    // make container the containing block instead of the viewport
    container.style.contain = "layout";
    popper = await mount(TestComp, container);
    const popBox2 = document.getElementById("popper").getBoundingClientRect();
    destroy(popper);

    // best positions are not the same relative to their containing block
    const [pos1, pos2] = positions;
    assert.equal(pos2.top, pos1.top + margin);
    assert.equal(pos2.left, pos1.left + margin);
    // best positions are the same relative to the viewport
    assert.equal(popBox1.top, popBox2.top);
    assert.equal(popBox1.left, popBox2.left);
});
```

File: addons/web/static/src/core/position_hook.js [hunk 5]
```diff
@@ -83,7 +83,9 @@ const DEFAULTS = {
  * @param {HTMLElement} reference
  * @param {HTMLElement} popper
  * @param {Options} options
- * @returns {PositioningSolution} the best positioning solution
+ * @returns {PositioningSolution} the best positioning solution, relative to
+ *                                the containing block of the popper.
+ *                                => can be applied to popper.style.(top|left)
```
Comment (brboi) [inline addons/web/static/src/core/position_hook.js:88]: not sure we need to update the JSDoc for a non exported function, but why not ;-)

File: addons/web/static/src/core/position_hook.js (hunk 2 above)
Comment (abd-msyukyu-odoo) [reply addons/web/static/src/core/position_hook.js:108]: Yeah it's not needed I made my change while thinking that `getBestPosition` could be exported but if we never plan to export it then we can use the fact that we reset to `0px` before calling it

**Review by brboi — APPROVED** [review]
Thanks for your work here ;-)

robodoo r+
